Initial Endpoint for Metrics API

echappen opened the pull request:
This API will serve metrics data to external services.

Metrics API consists of a Function App with an Azure HTTPTrigger event function for each endpoint.

The first endpoint added here is to fetch all Participant Uploads from the metrics database.

Piipan.Metrics.Models were extracted to be shared between the API and other metrics subsystems.

PiipanMetricsApi contains a few Serializers to help format the http responses.

For now, the API is using the openapi standard as a reference only.

kingcomma reviewed (commented):
Very cool! This is really coming together fast.

My comments are mostly limited to maintenance items. It would be great to resolve the documentation and access control items before merging.

Also, now that the individual projects are coming together in this sub-system, I think it's time to check repeatable package restores and updating `dependabot.yml` off the list, per the [source guidelines](https://github.com/18F/piipan/blob/main/docs/source-guidelines.md#guidelines-for-piipan-net-source-trees).

echappen commented on iac/create-metrics-resources.bash:
```diff
@@ -176,3 +176,52 @@ while IFS=, read -r abbr name ; do
 done < states.csv
+
+# Create Metrics API Function App in Azure
+METRICS_FUNC_APP_NAME=PiipanMetricsApi
```
I think I've resolved this in [this commit](https://github.com/18F/piipan/pull/325/commits/fbffcb9544d1d082173bae8d3e3b63503147ae5a)

echappen commented on iac/create-metrics-resources.bash:
```diff
@@ -176,3 +176,52 @@ while IFS=, read -r abbr name ; do
+echo "Waiting to publish api function app"
+sleep 60s
+
+echo "Configure settings on $METRICS_FUNC_APP_NAME"
+az functionapp config appsettings set \
+  --resource-group $RESOURCE_GROUP \
+  --name $METRICS_FUNC_APP_NAME \
+  --settings \
+    $DB_CONN_STR_KEY="$DB_CONN_STR" \
+  --output none
```
Thanks—sounds like it's a good time to switch this to an ARM template. I'll let you know when that's ready.

echappen replied:
@kingcomma Updating the api versions solved my ip formatting issues—thanks! I switched deployment of the API app to an [ARM template](https://github.com/18F/piipan/pull/325/commits/52e04b8fe618cebb490dac42ba5f35a4dc1eacda#) where I could force https, disable ftp, and restrict ips.

echappen commented on metrics/README.md:
```diff
@@ -0,0 +1,40 @@
+
+For now, there's a 1-1 relationship between a specific metric needing to be captured and a function within this app.
+
+### Metrics API
+
+The Metrics API is a Function App made up of Azure HTTPTrigger event functions.
+
+For now, there's a 1-1 relationship between an API endpoint and a function within this app.
```
Done!

echappen commented:
@kingcomma Ok I think I've resolved all of the items you pointed out:
- Locked down Metrics API to GSA ip's by switching to an ARM template; also forces HTTPS and disables FTP
- Added package lockfiles to all metrics projects
- Added metrics packages to Dependabot
- Added documentation about database connection strings

I added pagination logic to [GetParticipantUploads.cs](https://github.com/18F/piipan/pull/325/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.csproj&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.yaml&file-filters%5B%5D=dotfile#diff-fe5ef8d670e742ef5b1096de65fd94d9c63160b32a8f89177bd48e25e0bde483R22) since you've last looked at this. The `Run` function is the entry point for the api endpoint. This class has gotten pretty large — if you have any thoughts on breaking this into smaller service classes, I'll take them. But if you think I should move on and revisit/refactor at a later time, that's ok with me too.

kingcomma commented on iac/create-metrics-resources.bash:
```diff
@@ -91,9 +91,14 @@ psql -U $DB_ADMIN_NAME@$DB_SERVER_NAME -p 5432 -d $DB_NAME -w -v ON_ERROR_STOP=1
 
 ### Function App stuff
-FUNC_APP_NAME=PiipanMetricsFunctions
+FUNCTIONS_UNIQ_STR=`az deployment group create \
+  --resource-group $RESOURCE_GROUP \
+  --template-file ./arm-templates/unique-string.json \
+  --query properties.outputs.uniqueString.value \
+  -o tsv`
+FUNC_APP_NAME=PiipanMetricsFunctions${FUNCTIONS_UNIQ_STR}
+FUNC_STORAGE_NAME=piipanmetricsstorage${FUNCTIONS_UNIQ_STR}
```
**Suggestion**: Now that you're leveraging ARM templates, I think you can cut this out and then use the `uniquestring(resourceGroup().id)` template function you're using in `metrics-api.json` to compose any strings that still rely on `FUNCTIONS_UNIQ_STR`. I think IaC refactoring is on the agenda in the next few sprints, so I don't think this is strictly necessary right now.

(Though note that any strings you'll need elsewhere in the script, like `METRICS_FUNC_APP_NAME` would need to be captured as output from the call to `az deployment group create`).

echappen replied:
Thanks, I've done this for metrics-api.json. Once I put the other function app into an ARM template (I'll do this is another PR), I can remove the variable entirely.

kingcomma commented on iac/create-metrics-resources.bash:
```diff
@@ -176,3 +181,51 @@ while IFS=, read -r abbr name ; do
+      resourceTags="$RESOURCE_TAGS" \
+      location=$LOCATION
+
+echo "Configure settings on $METRICS_FUNC_APP_NAME"
+az functionapp config appsettings set \
+  --resource-group $RESOURCE_GROUP \
+  --name $METRICS_FUNC_APP_NAME \
+  --settings \
+    $DB_CONN_STR_KEY="$DB_CONN_STR" \
+  --output none
```
**Suggestion**: This setting could be passed to `metrics-api.json` and configured under `properties. siteConfig`. Having all the configuration in one place might make it easier to manage and may be necessary down the road if/when we move towards ARM templates as much as possible.

echappen replied:
Done!

kingcomma reviewed (approved):
👍

I put a couple ARM template suggestions below that don't need to hold this up, but will likely be helpful to address in the near future and/or when we refactor the IaC script.

As for the class size, I think there are a few methods that would be natural candidates for pulling out into a helper class (e.g., `StrToIntWithDefault` and `NextPageParams`) as the API expands to include other metrics. We haven't defined those metrics yet, so I think it's up to you when it makes sense.

And as you know `ConnectionString` will naturally be abstracted out with #192.

echappen commented:
@kingcomma Thank you for your thorough review here! I've incorporated some more setup into the `metrics-api.json` template, and created an issue to finish converting the rest of the setup script to ARM template(s). 

As for breaking up `GetParticipantUploads` into helper classes, I agree with your natural candidates and think adding a new metric endpoint will help guide me on that. In the mean time, all public methods in that class are tested. 

Now that code climate is green I'm going to merge so I can move on to showing results in the dashboard for our next demo. Thanks again!
